s3transfer/processpool.py

```diff
@@ -0,0 +1,193 @@
+
+
+class GetObjectWorker(multiprocessing.Process):
+    _MAX_ATTEMPTS = 5
+    _IO_CHUNKSIZE = 2 * MB
+
+    def __init__(self, queue, client_factory, transfer_monitor, osutil):
+        """Fulfills GetObjectJobs
+
+        Downloads the S3 object, writes it to the specified file, and
+        renames the file to its final location if it completes the final
+        job for a particular transfer.
+
+        :param queue: Queue for retrieving GetObjectJob's
+        :param client_factory: ClientFactory for creating S3 clients
+        :param transfer_monitor: Monitor for notifying
+        :param osutil: OSUtils object to use for os-related behavior when
+            performing the transfer.
+        """
+        super(GetObjectWorker, self).__init__()
+        self._queue = queue
+        self._client_factory = client_factory
+        self._transfer_monitor = transfer_monitor
+        self._osutil = osutil
+        self._client = None
+
+    def run(self):
+        # Client are not pickleable so their instantiation cannot happen
+        # in the __init__ for processes that are created under the
+        # spawn method.
+        self._client = self._client_factory.create_client()
+        while True:
+            job = self._queue.get()
+            if job == SHUTDOWN_SIGNAL:
+                return
+            if not self._transfer_monitor.get_exception(job.transfer_id):
+                self._run_get_object_job(job)
+            remaining = self._transfer_monitor.notify_job_complete(
+                job.transfer_id)
+            if not remaining:
+                self._finalize_download(
+                    job.transfer_id, job.filename, job.final_filename
+                )
+
+    def _run_get_object_job(self, job):
+        try:
+            self._do_get_object(
+                bucket=job.bucket, key=job.key, filename=job.filename,
+                extra_args=job.extra_args, offset=job.offset
+            )
+        except Exception as e:
+            self._transfer_monitor.notify_exception(job.transfer_id, e)
+
+    def _do_get_object(self, bucket, key, extra_args, filename, offset):
+        last_exception = None
+        for i in range(self._MAX_ATTEMPTS):
+            try:
+                response = self._client.get_object(
+                    Bucket=bucket, Key=key, **extra_args)
+                self._write_to_file(filename, offset, response['Body'])
+                return
+            except S3_RETRYABLE_DOWNLOAD_ERRORS as e:
+                logger.debug('Retrying exception caught (%s), '
+                             'retrying request, (attempt %s / %s)', e, i,
+                             self._MAX_ATTEMPTS, exc_info=True)
+                last_exception = e
+        raise RetriesExceededError(last_exception)
+
+    def _write_to_file(self, filename, offset, body):
+        with open(filename, 'rb+') as f:
+            f.seek(offset)
+            chunks = iter(lambda: body.read(self._IO_CHUNKSIZE), b'')
```

Comment:
It is the best value that I found testing it out in my environment. I think it would make sense to open it up in the future if we determine its something people want to tweak. I can put a TODO about potentially moving it to configuration.